compiler/rustc_middle/src/mir/interpret/mod.rs

```diff
@@ -428,26 +434,33 @@ impl<'tcx> TyCtxt<'tcx> {
     /// Make sure to call `set_alloc_id_memory` or `set_alloc_id_same_memory` before returning such
     /// an `AllocId` from a query.
     pub fn reserve_alloc_id(self) -> AllocId {
-        self.alloc_map.lock().reserve()
+        self.alloc_map.reserve()
     }
 
     /// Reserves a new ID *if* this allocation has not been dedup-reserved before.
     /// Should not be used for mutable memory.
     fn reserve_and_set_dedup(self, alloc: GlobalAlloc<'tcx>, salt: usize) -> AllocId {
-        let mut alloc_map = self.alloc_map.lock();
         if let GlobalAlloc::Memory(mem) = alloc {
             if mem.inner().mutability.is_mut() {
                 bug!("trying to dedup-reserve mutable memory");
             }
         }
         let alloc_salt = (alloc, salt);
-        if let Some(&alloc_id) = alloc_map.dedup.get(&alloc_salt) {
+        let mut dedup = self.alloc_map.dedup.lock();
+        if let Some(&alloc_id) = dedup.get(&alloc_salt) {
             return alloc_id;
         }
-        let id = alloc_map.reserve();
+        let id = self.alloc_map.reserve();
         debug!("creating alloc {:?} with id {id:?}", alloc_salt.0);
-        alloc_map.alloc_map.insert(id, alloc_salt.0.clone());
-        alloc_map.dedup.insert(alloc_salt, id);
+        // We just reserved, so should always be unique.
+        assert!(
```

Comment:
I find code that has side-effects inside assertions pretty hard to follow. Is there a good way to avoid that?

Reply:
I pulled the assert out to a separate line - does that help?

Reply:
Yes, that's better, thanks!

Looking at the API, we have `try_insert`. It's still unstable but might be worth it? Then it can be a single line.

Reply:
I'm not a fan of adding unstable API usage given that we don't care about the semantic here (try_insert is useful primarily in that it leaves the entry untouched, rather than overwriting it).